### PR TITLE
Kmeans

### PR DESCRIPTION
**Pull request by rmjarvis**
A common usage of TreeCorr is to split up a large survey into multiple smaller patches for the purpose of jackknife or bootstrap error estimation.  The typical algorithm for finding these patches is [k-means clustering](https://en.wikipedia.org/wiki/K-means_clustering).  

As the first step toward incorporating the jackknife calculation into TreeCorr to be done natively (Issue #24), I wanted to implement a clustering algorithm, so it could find a good set of patches for you.  So I investigated the existing state of the art for doing k-means in python.  I found the following implementations:
* [scipy.cluster.vq.kmeans](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.vq.kmeans.html)
* [scipy.cluster.vq.kmeans2](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.vq.kmeans2.html#scipy.cluster.vq.kmeans2)
* [kmeans_radec](https://github.com/esheldon/kmeans_radec)
* [pyclustering.cluster.kmeans](https://pyclustering.github.io/docs/0.8.2/html/da/d97/namespacepyclustering_1_1cluster_1_1kmeans.html)
* [sklearn.cluster.KMeans](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html#sklearn.cluster.KMeans)
* [sklearn.cluster.MiniBatchKMeans](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.MiniBatchKMeans.html#sklearn.cluster.MiniBatchKMeans)

I made a [notebook](https://github.com/rmjarvis/TreeCorr/blob/kmeans/devel/kmeans.ipynb) comparing the different algorithms using a random million galaxies from the DES SV footprint (chosen because it is a real-life use case that has some ratty edges to deal with).

The ideal patches would be essentially uniform in size according to some measure of the effective area of the patch.  The "area" of a set of points isn't completely well defined, but I picked the "inertia" as a decent proxy for what we care about.  The inertia is defined as the sum of the distance squared of each point in a patch to the center of the patch.  I_k = Sum_i d_ik^2.

The k-means algorithm is designed to minimize the total inertia of all the patches added together.  However, this isn't directly what we want.  We want the sizes to be all close to equal, not necessarily have the minimum total (or mean).  I.e. we want the rms inertia to be small.  Fortunately, the process of minimizing the total inertia does tend to select patches with small rms as well, but it is worth noting that this is not directly targeted by the normal k-means algorithm.  And furthermore, the k-means algorithm almost never finds the true global minimum inertia.  The quality of the local minimum depends pretty strongly on the choice of initial centers to seed the iterative part of the algorithm.

Comparing the results of the various k-means implementations, I found that they all tend to be either fairly slow, taking a minute or more for just 1 million objects, or they have very high rms inertia.
I reran each code multiple times using a different random million selected from the original catalog (of around 16 million objects).  Here is a scatter plot of the time vs rms inertia for the various codes.  
image

Since there was no existing implementation I was particularly happy with, I implemented it myself.  It turns out (not surprisingly) that the ball tree structure that TreeCorr already uses for doing correlation functions quickly is also very useful for doing k-means quickly.  Also, the quality of the k-means result is pretty dependent on the choice of the initial centers, and using the ball tree for the initialization produces reliably better results than the initialization methods used by other packages.

The big red dots in the lower left corner are the TreeCorr implementation of the standard k-means clustering algorithm.  It typically takes about 1 or 2 seconds to classify these 1 million points into 40 patches, and the rms inertia is usually less than any other implementation.  The [notebook](https://github.com/rmjarvis/TreeCorr/blob/kmeans/devel/kmeans.ipynb) also includes plots of total inertia, rms size using according to the *mean* d^2 rather than sum, and rms counts.  The TreeCorr algorithm tends to be the best k-means implementation according to any of these metrics.

In addition, you can see some slightly smaller orange dots, which have even lower rms inertia but take very slightly longer to run.  These are an alternate algorithm I developed based on k-means, which penalizes patches with a larger-than-average inertia, so they give up some of their outer points to patches with smaller inertia.  In other words, it explicitly targets making the rms inertia as small as possible.  But in practice, it is not much worse in terms of total inertia, which is the thing that the standard k-means algorithm technically minimizes.

So far, the API for the standard k-means algorithm is to make an NField from a catalog and then run `field.run_kmeans(npatch)`.  However, in future PRs, I'll make this directly callable from a Catalog object and add some functionality to actually use the patches in the correlation function processing.

The alternate algorithm is available as `field.run_kmeans(npatch, alt=True)`.  I left this as a non-default option for two reasons.  First, it's not actually the real k-means, so I didn't want to confuse people who just want to use this for regular k-means clustering.  But second, I'm not completely sure that it is always stable.  There is a free parameter in the penalty function I chose, which I set to 3.  Setting it to 4 gave even better results (slightly), but at 5 the algorithm broke down with neighboring patches trading escalating numbers of points between each other until one of them had no points left.  

I couldn't convince myself that 4 was actually a magic number and not just the particular value for this data set.  So 3 might be safe, or there might be data sets where that also leads to this runaway trading failure mode.  I know the regular k-means algorithm can't get into this mode, so it's always safe.  Therefore, I think it's better to force the user to intentionally select the alternate algorithm if they really care about having a low rms patch size, with the normal algorithm being the backup if the alternate one fails for them.

Pinging possibly interested people: @esheldon @ebaxter @matroxel @danielgruen @seccolf @NiallMac @carlessanchezalonso @xuod

### CODE REVIEW

**Comment by esheldon** [comment]
Looks good mike.  

The interesting feature for my `kmeans_radec` code is that it works on the sphere, giving roughly equal area with the right metric.   I think yours also has that most important feature.

I never did put much effort into optimizing `kmeans_radec`, figuring people run this sort of code rarely so speed isn't critical (I did start a `method='fast'` version but n ever finished it).

I think the improvement in scatter is the most interesting part here.  It might have a real impact on the science.

**Comment by rmjarvis** [comment]
> The interesting feature for my kmeans_radec code is that it works on the sphere, giving roughly equal area with the right metric. I think yours also has that most important feature.

Yes, that's right.  For the other codes, I pass them x,y,z on the unit sphere, rather than ra, dec values.  But giving ra, dec directly would give wrong results.
